### DISCUSSION TITLE
Package Architecture

### DISCUSSION THREAD

**Discussion by TheGreatAxios** [Dirt-Road-Development/skale.js#54] [Ideas]
The goal is is offering flexibility at all points in the "process" and enabling development on SKALE in a variety of ways; i.e graphical, CLI, via servers, or on a client (non-graphical).

Take **File Storage** as an example.

### Abstraction & Utils
1. If you want to utilize the contract by calling the ABI + Address (from constants) and setup your own instance using Web3.js instead of Ethers.
2. If you want to utilize the wrapper you can create a new instance of Filestorage to unlock all of the core functions on the smart contract without having to manage the boilerplate of doing so yourself
3. If you want greater abstraction that is also included under Filestorage under the Utils sections (i.e similar to Filestorage.js)

1-3 Moving up from 1 to 3, we are increasing the level of abstraction enabling different usages by different developers. Most developers may lean on one or the other, but, by providing all of them we won't deter those who want to start at the bottom from building; additionally, maybe newer developers are a bit wary of components being too opinionated as it may not match up with tutorials they are following. The goal is to have example use cases for each and every example showcasing usage allowing any developer to quickly get building on SKALE.

### UI

The next step in the process is creating components that are flexible. The original SKALE Chain UI found here (https://github.com/Dirt-Road-Development/schain-dashboard) is a good example of why modular building will work here. 

Each and every one of these pages is small enough in scope to be it's own mini "dApp" or "frontend".  Ideally, each and every package should have all of the logic from 1-3 above wrapped into components that match a specific style and brand guideline while being as customizable as possible. This will allow projects to drop components into their dApps to facilitate actions specific to SKALE.

Core UI packages should be exported from a skale-ui package that is an aggregation of all ui packages. Additionally, for higher "complexity" UI; a unified frontend can be created i.e filestorage-dapp which contains a working dApp of all components. 

### CLI

Command Line Interfaces (CLI) are one of the fastest tools that developers can use to get started and facilitate actions. In this case, I believe it is in our best interest to create a unified CLI package that can be ran as ```npx skale``` or something similiar that then has access to all of the existing functional packages. If every package were to have their own CLI, it would result in developers having to install and manage all of those dependencies and tools. 

Building on SKALE should be simple. A single command should install the CLI and have a walkthrough for setup. From there, it should be dynamic in where it pulls data from, but also, be very declarative with steps to enable even non CLI users to quickly and easily read/write data to/from SKALE Chains.

**Comment by adiled** [comment]
A single-entry CLI is a very good idea.

On the abstraction and util layer, relating to my proposition @ https://github.com/Dirt-Road-Development/skale.js/issues/16#issuecomment-1357202200 and I'll quote the effective gist of it:

> A practical scenario I'd propose, that may spread bliss across the dev tooling through all verticals:
> 
> 1. Contract wrappers are two-pronged, catering both user classes on even keel: ship events as an API, separately from method callers (`Contract.ts/events`).
> 2. A layer goes between primitive vertical and client verticals: bind the wrapper prongs by associating events and methods (`Contract.ts/context`) AKA a spoon.
> 3. SDKs consume the union above, with a provision to being watchful of state. e.g. `filestorage`, `ima` becoming inclusive for reactionary devs. Initially, this may be under an experimental flag.

The first 2 points are my suggestion for L2 abstraction, and it'll be best to spec it out exclusively, but here I'll contextualize the motivation a bit, particularly highlighting the reactionary dev, which I believe is still deterred from DApp development, despite so much learning content.

The need for feedback following transaction initiation arises when one raises the bar for UX. Why this problem is not yet pronounced is because DApp ecosystem as a whole has only now barely tapped on the surface of utility, granted layer 2, and generally the larger UX community has been reluctant to engage.

By example, when you build a file manager beyond an MVP atop decentralized file system, any write operation involving contract interaction, seeks beyond transactional feedback. Now these are indeed fuzzy problems because naturally we are dealing with an underlying probabilistic system, but on-ground, for end-usability, you are hungry for state.
How I fed the hunger of `filestorage-ui` within [`defilemanager.ts`](https://github.com/skalenetwork/filestorage-ui/blob/main/src/packages/filemanager/filemanager.ts) was by playing inception on observables which were being used for sequencing transactions, and then exposing an event bus in addition to prior imperative API. The contract events in the mix would close the UX loop, and of course there are more than few ways to implement this. I haven't yet worked with IMA but I imagine it is principally a lot of bridging, which again at user-end should be state hungry for best of UX.

How I imagine this materializes within the package structure is within L2, a very thin event layer with subscribability can enable an events interface and one more which can be thought of as L2.5. Depending on the scope of L3s, this could be leveraged within them, but the use-cases are far-reaching.

On the UI side, I'd say amen to micro/modular-frontend strategy, there's a lot to explore there, and I have all the support for everything frontend that shakes the ecosystem out of lock-in to a particular stack.

**Reply by TheGreatAxios** [reply]
Thanks for such a detailed response, I think we are aligned well here. 

To clarify as you bring up a number of good topics: 

1)  ALL **WRITE** operations should result in a response of two types: Success or Error
2) These types allow for a UI or other tools to have a standardized approach to handling the event result of a write operation
3) This can be seen in your queue module underneath defilemanager.ts (link above)

If I am correct in this clarification, would you be open to building these process as a utility under the utils package? I am interested to see if the state and results can be defined package by package and passed into this event handler which can then take dynamic information and handle the result properly.

This would then be used in the context layer. What I am saying is I see three layers:

1) The Contract Wrapper i.e contract.ts
2) The Event Layer i.e events.ts -> Defined package-by-package 
3) The Context Layer -> Contracts.ts + events.ts adhering to your approach above

---

@adiled I will create a more detailed contributor doc that is related to package creation and architecture. I will share the PR here once created so we can discuss how to properly structure all of these packages in greater detail to adhere to this very modular approach. Let me know if you have any other thoughts/ideas on this or what is being confirmed above.

**Reply by adiled** [reply]
:+1:  That's well summarized. With some exploration I'll be creating a spec document for this, identifying possible techniques and feasibility across cases, and then derive reference implementations.
